### ISSUE TITLE
pubsub: Publish RPC traces not not in the same tree

### ISSUE THREAD

**Issue by fastest963** [googleapis/google-cloud-go#1347]
## Client

pubsub

## Describe Your Environment

CentOS 7 on GCE

## Expected Behavior

I expected the `google.pubsub.v1.Publisher.Publish` traces to use the span off the context so it's in the same tree.

## Actual Behavior

We're seeing in stackdriver trace `google.pubsub.v1.Publisher.Publish` as the only span in a tree.

We see:
![image](https://user-images.githubusercontent.com/112555/53969537-f931bf00-40c6-11e9-9472-7d0f49863c0d.png)

but I expected it to be a span in this tree:
![image](https://user-images.githubusercontent.com/112555/53969612-1c5c6e80-40c7-11e9-940c-6b67f739fa02.png)

Is this because the submits are batched and so the original context is lost? I'm not familiar enough with the tracing to know if this is expected or not.

**Comment by jadekler** [comment]
> Is this because the submits are batched and so the original context is lost?

That's exactly it, yes. Well, that and that the batching happens asynchronously.

I wonder whether we can hold onto the context. I'll mark this a feature request.

**Comment by ethiery** [comment]
I have created a [minimal reproducible example](https://github.com/ethiery/MRE_opencensus_pubsub) for this.

@fastest963  it includes an example of an alternative that might or might not suit your needs. The [Go Cloud Development Kit project](https://gocloud.dev) includes a GCP PubSub client which nests spans as you expect.

@jadekler for the record I am also interested in this "feature".

**Comment by odeke-em** [comment]
Thank you for filing this issue @fastest963 and @ethiery for the vote, @jadekler for the answers!

As @jadekler has reported, in deed the problem is because we are processing/bundling messages but discarding information about who created us. OpenCensus has a concept of a Span.Link https://opencensus.io/tracing/span/link/#visuals
and because we are batching requests to aid in both quota maximization as well as preserving network bandwidth and resources, we can't remove batching.

@ethiery while empirically that looks "correct", the Go CDK just creates a span that is ended after a span has been added to the bundler as per:
* Span Start and End: https://github.com/google/go-cloud/blob/dc644529bb532ebfe059c8ae23fcc9fda49d1791/pubsub/pubsub.go#L246-L247
* Adding to bundler: https://github.com/google/go-cloud/blob/dc644529bb532ebfe059c8ae23fcc9fda49d1791/pubsub/pubsub.go#L272

Your screenshot in
![](https://github.com/ethiery/MRE_opencensus_pubsub/raw/master/gocdk_main_trace.png)
is only made of the screenshot that performs batch processing and suffers from the same problem reported here. Just that this repo doesn't create a span when adding items to the batch.

Thank you for that screenshot @ethiery , it reveals a bit of a problem btw, that the Go CDK's bundler takes 312ms to just add an item to an in-process bundler for batch processing, that to me looks like a lot of blocking that we should report to them -- kindly paging @zombiezen @jba @vangent, perhaps we can open this issue separately on the Go CDK project.

### Remedy

Basically to handle asynchronous process, when batching, we'll extract a parent link and when publishing requests create a span retaining the parent common link -- basically batching by parent link.

**Comment by vangent** [comment]
> the Go CDK's bundler takes 312ms to just add an item to an in-process bundler for batch processing

That trace means that it took 312ms to send the message; `batcher.Add` waits until the message is actually sent, possibly batched with other messages (there's a separate `AddNoWait` function). 312ms seems reasonable for an RPC to GCP Pub/Sub.

**Comment by vangent** [comment]
> the Go CDK just creates a span that is ended after a span has been added to the bundler

So this comment is not correct; the span is ended after the message has actually been **sent** by the bundler, not just Added and sent asynchronously.

**Comment by odeke-em** [comment]
>waits until the message is actually sent, possibly batched with other messages (there's a separate AddNoWait function). 312ms seems reasonable for an RPC to GCP Pub/Sub.

Interesting! My apologies for the misinterpretation, I saw `AddNoWait` https://godoc.org/gocloud.dev/internal/batcher#Batcher.AddNoWait and used that to infer my response about the asynchronity.
312ms definitely seems reasonable for a synchronous RPC. 

Alright, so the two libraries packages don't batch responses the same way, this one batches purely asynchronously, the Go CDK batches synchronously.

I'll go on with the remedy about linking, what do y'all think @jadekler @broady?

**Comment by jadekler** [comment]
After giving this some more thought, I don't think we should do this. Providing a ctx to Publish that gets used in Publish RPCs - while _maybe_ a good thing for spans - has confusing semantics for cancellation and expiration. Does cancelling the context passed to a complete Publish cause the yet-to-complete Publish RPC to fail?

It's also not clear to me what the spans should look like. It's not a tree with obvious parents: multiple Publish calls (the "parent") can feed into a single Publish RPC (the "child"), which would be confusing to interpret on a chart.

If someone can present a clear case of how this would all work I'm happy to explore, but for now I'm going to close this issue.
